runtime/lib/mirrors_impl.dart

```diff
@@ -1664,15 +1664,35 @@ class _Mirrors {
     return classMirror;
   }
 
-  static TypeMirror reflectType(Type key) {
-    var typeMirror = _instantiationCache[key];
-    if (typeMirror == null) {
-      typeMirror = makeLocalTypeMirror(key);
-      _instantiationCache[key] = typeMirror;
-      if (typeMirror is ClassMirror && !typeMirror._isGeneric) {
-        _declarationCache[key] = typeMirror;
+  static TypeMirror reflectType(Type key, [Iterable<Type> typeArguments]) {
+    if (typeArguments == null || typeArguments.isEmpty) {
+      var typeMirror = _instantiationCache[key];
+      if (typeMirror == null) {
+        typeMirror = makeLocalTypeMirror(key, arguments);
+        _instantiationCache[key] = typeMirror;
+        if (typeMirror is ClassMirror && !typeMirror._isGeneric) {
+          _declarationCache[key] = typeMirror;
+        }
       }
+      return typeMirror;
+    } else {
+      // We can't cache type provided in `key` since actual type will depend
```

Comment:
I'm not sure I understand the comment. There is no caching in the line that follows.
Also it actually *does* cache the type eventually (line 1689)